rqt_py_common/src/rqt_py_common/topic_helpers.py

```diff
@@ -28,161 +28,167 @@
-        if type_name == 'string':
-            return str
-        elif type_name == 'bool':
-            return bool
-        else:
-            return type(roslib.msgs._convert_val(type_name, 0))
-    else:
-        return roslib.message.get_message_class(type_name)
+from rclpy import logging
 
 
-def get_field_type(topic_name):
+def get_topic_names_and_types(node=None):
     """
-    Get the Python type of a specific field in the given registered topic.
-    If the field is an array, the type of the array's values are returned and the is_array flag is set to True.
-    This is a static type check, so it works for unpublished topics and with empty arrays.
+    Get avaliable topic names and types.
 
-    :param topic_name: name of field of a registered topic, ``str``, i.e. '/rosout/file'
-    :returns: field_type, is_array
+    Note: in ROS2, only nodes can query the topic information
+    @param node: a ROS node
+    @type node: rclpy.node.Node or None
+
+    If node is None, then this method will create a node, use it to get topic
+    information and then destroy the node.
     """
-    # get topic_type and message_evaluator
-    topic_type, real_topic_name, _ = get_topic_type(topic_name)
-    if topic_type is None:
-        # qDebug('topic_helpers.get_field_type(%s): get_topic_type failed' % (topic_name))
-        return None, False
+    if node is not None:
+        return node.get_topic_names_and_types()
 
-    message_class = roslib.message.get_message_class(topic_type)
-    if message_class is None:
-        qDebug('topic_helpers.get_field_type(%s): get_message_class(%s) failed' %
-               (topic_name, topic_type))
-        return None, False
+    import rclpy
+    shutdown_rclpy = False
+    if not rclpy.ok():
+        shutdown_rclpy = True
+        rclpy.init()
 
-    slot_path = topic_name[len(real_topic_name):]
-    return get_slot_type(message_class, slot_path)
+    node = rclpy.create_node("TopicHelpers__get_topic_names_and_types")
 
+    # Give the node time to learn about the graph
+    rclpy.spin_once(node, timeout_sec=0.05)
 
-def get_slot_type(message_class, slot_path):
-    """
-    Get the Python type of a specific slot in the given message class.
-    If the field is an array, the type of the array's values are returned and the is_array flag is set to True.
-    This is a static type check, so it works for unpublished topics and with empty arrays.
+    topic_list = node.get_topic_names_and_types()
 
-    :param message_class: message class type, ``type``, usually inherits from genpy.message.Message
-    :param slot_path: path to the slot inside the message class, ``str``, i.e. 'header/seq'
-    :returns: field_type, is_array
-    """
-    is_array = False
-    fields = [f for f in slot_path.split('/') if f]
-    for field_name in fields:
+    node.destroy_node()
+    if shutdown_rclpy:
+        rclpy.shutdown()
+
+    return topic_list
+
+
+_message_class_cache = {}
+def get_message_class(message_type):
```

Comment:
The logic in this function is certainly going to break with the upcoming IDL changes.

Reply:
Definitely ping me when the upcoming idl changes are solidified